plugin/plugins/telemetry/telemetry.go

```diff
@@ -0,0 +1,521 @@
+package telemetry
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"path"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	logging "github.com/ipfs/go-log/v2"
+	ipfs "github.com/ipfs/kubo"
+	"github.com/ipfs/kubo/config"
+	"github.com/ipfs/kubo/core"
+	"github.com/ipfs/kubo/core/corerepo"
+	"github.com/ipfs/kubo/plugin"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/pnet"
+	multiaddr "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+var log = logging.Logger("telemetry")
+
+const (
+	modeEnvVar   = "IPFS_TELEMETRY_PLUGIN_MODE"
+	uuidFilename = "telemetry_uuid"
+	endpoint     = "http://127.0.0.1:8083"
+)
+
+type pluginMode int
+
+const (
+	modeInfo pluginMode = iota
+	modeOptIn
+	modeOptOut
+)
+
+// repoSizeBuckets defines size thresholds for categorizing repository sizes.
+// Each value represents the upper limit of a bucket in bytes (except the last)
+var repoSizeBuckets = []uint64{
+	1 << 30,   // 1 GB
+	5 << 30,   // 5 GB
+	10 << 30,  // 10 GB
+	100 << 30, // 100 GB
+	500 << 30, // 500 GB
+	1 << 40,   // 1 TB
+	10 << 40,  // 10 TB
+	11 << 40,  // + anything more than 10TB falls here.
+}
+
+var uptimeBuckets = []time.Duration{
+	1 * 24 * time.Hour,
+	24 * time.Hour,
+	3 * 24 * time.Hour,
+	7 * 24 * time.Hour,
+	14 * 24 * time.Hour,
+	30 * 24 * time.Hour,
+	31 * 24 * time.Hour, // + anything more than 30 days falls here.
+}
+
+// A LogEvent is the object sent to the telemetry endpoint.
+type LogEvent struct {
+	UUID *string `json:"uuid"`
+
+	AgentVersion *string `json:"agent_version"`
+
+	PrivateNetwork *bool `json:"private_network"`
+
+	BootstrappersCustom *bool `json:"bootstrappers_custom"`
+
+	RepoSizeBucket *uint64 `json:"repo_size_bucket"`
+
+	UptimeBucket *time.Duration `json:"uptime_bucket"`
+
+	ReproviderStrategy *string `json:"reprovider_strategy"`
+
+	RoutingType                 *string `json:"routing_type"`
+	RoutingAcceleratedDHTClient *bool   `json:"routing_accelerated_dht_client"`
+	RoutingDelegatedCount       *int    `json:"routing_delegated_count"`
+
+	AutoNATServiceMode  *string `json:"autonat_service_mode"`
+	AutoNATReachability *string `json:"autonat_reachability"`
+
+	SwarmEnableHolePunching  *bool `json:"swarm_enable_hole_punching"`
+	SwarmCircuitAddresses    *bool `json:"swarm_circuit_addresses"`
+	SwarmIPv4PublicAddresses *bool `json:"swarm_ipv4_public_addresses"`
+	SwarmIPv6PublicAddresses *bool `json:"swarm_ipv6_public_addresses"`
+
+	AutoTLSAutoWSS            *bool `json:"auto_tls_auto_wss"`
+	AutoTLSDomainSuffixCustom *bool `json:"auto_tls_domain_suffix_custom"`
+
+	DiscoveryMDNSEnabled *bool `json:"discovery_mdns_enabled"`
+
+	PlatformOS            *string `json:"platform_os"`
+	PlatformArch          *string `json:"platform_arch"`
+	PlatformContainerized *bool   `json:"platform_containerized"`
+}
```

Comment:
It's all pointers so that we can distinguish between set and unset things. Particularly in the future when we add more fields and previous versions might not have them.

That said, it is a pain to work with these pointers and a panic footgun. So I think I will switch back unless someone has a good reason to leave things like this.

Reply:
I think that the zero value (unset) means the same as having a `nil` pointer for all of these. In other words I do not see a case where we need to differentiate between unset and the zero value.  So, I prefer non-pointer.